### PR TITLE
Add find_by

### PR DESCRIPTION
**Pull request by fdhadzh**

### CODE REVIEW

File: lib/no_brainer/criteria/find_by.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module NoBrainer::Criteria::FindBy
+  extend ActiveSupport::Concern
+
+  def find_by(*block)
+    where(*block).first
```
Comment (nviennot) [inline lib/no_brainer/criteria/find_by.rb]: use `(*args, &block)` instead of `(*block)`

**Comment by nviennot** [comment]
rename `find_by` -> `find_by?` and `find_by!` -> `find_by`. So we are consistent with the rest of the API (`find` raises, `find?` doesn't)

That's it :)
And thank you :)

**Comment by fdhadzh** [comment]
Success

**Comment by nviennot** [comment]
The current implementation of `find` is:

> * `Model.find(id)` is equivalent to `Model.unscoped.where(:id => id).first!`.
> * `Model.find?(id)` is equivalent to `Model.unscoped.where(:id => id).first`.
>
> Note that `find()` is only defined on the `Model` class, and not on criteria.

I'd like to have `find` and `find_by` semantics to be similar. So I might have to remove the `unscoped` semantic. I'll sleep on it. If you have any opinion on what should be done, and why, I'd be happy to hear it.

**Comment by fdhadzh** [comment]
I can add find method to criteria or something another?

**Comment by nviennot** [comment]
I'll deal with it.

**Comment by fdhadzh** [comment]
spzj1uqdwdq

**Comment by nviennot** [comment]
Currenty, `find` is implemented with RQL `get()`. To do this, we can't use the `default_scope` or apply `find` on a criteria.
To use `find` on a criteria, it means that we need to use `get_all()`. However, from my benchmark, it appears that it's 50% slower.
So that mean that to have `find` to match the semantics of `find_by`, we'll have to give up 50% of the speed of the fast path `Model.find(id)`. I'm not so thrilled about it :/

**Comment by nviennot** [comment]
hahahaha (seeing the image :D)

**Comment by nviennot** [comment]
Merged it in, changed the behavior of `find()`.

**Comment by kuldeepaggarwal** [comment]
It is inconsistence with Rails ActiveRecord API, I think we should develop the public API as same like Rails.

**Comment by nviennot** [comment]
We settled on removing it: https://github.com/nviennot/nobrainer/issues/145

`find_by()` does not raise with AR, but `find()` does. I'm not too happy about introducing inconsistent API.

**Pull request closed without merging**